onchain/rollups/contracts/consensus/authority/AuthorityHistoryPairFactory.sol

```diff
@@ -0,0 +1,112 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+pragma solidity ^0.8.8;
+
+import {IAuthorityHistoryPairFactory} from "./IAuthorityHistoryPairFactory.sol";
+import {Authority} from "./Authority.sol";
+import {IAuthorityFactory} from "./IAuthorityFactory.sol";
+import {History} from "../../history/History.sol";
+import {IHistoryFactory} from "../../history/IHistoryFactory.sol";
+
+/// @title Authority-History Pair Factory
+/// @notice Allows anyone to reliably deploy a new Authority-History pair.
+contract AuthorityHistoryPairFactory is IAuthorityHistoryPairFactory {
+    IAuthorityFactory immutable authorityFactory;
+    IHistoryFactory immutable historyFactory;
+
+    /// @notice Constructs the factory.
+    /// @param _authorityFactory The `Authority` factory
+    /// @param _historyFactory The `History` factory
+    constructor(
+        IAuthorityFactory _authorityFactory,
+        IHistoryFactory _historyFactory
+    ) {
+        authorityFactory = _authorityFactory;
+        historyFactory = _historyFactory;
+
+        emit AuthorityHistoryPairFactoryCreated(
+            _authorityFactory,
+            _historyFactory
+        );
+    }
+
+    function getAuthorityFactory()
+        external
+        view
+        override
+        returns (IAuthorityFactory)
+    {
+        return authorityFactory;
+    }
+
+    function getHistoryFactory()
+        external
+        view
+        override
+        returns (IHistoryFactory)
+    {
+        return historyFactory;
+    }
+
+    function newAuthorityHistoryPair(
+        address _authorityOwner
+    ) external override returns (Authority authority_, History history_) {
+        authority_ = authorityFactory.newAuthority(address(this));
+        history_ = historyFactory.newHistory(address(authority_));
+
+        authority_.setHistory(history_);
+        authority_.transferOwnership(_authorityOwner);
+    }
+
+    function newAuthorityHistoryPair(
+        address _authorityOwner,
+        bytes32 _salt
+    ) external override returns (Authority authority_, History history_) {
+        _salt = calculateCompoundSalt(_authorityOwner, _salt);
```

Comment:
Just curious why not directly pass in the "compound salt" as the function parameter `_salt`? Which means the computation is done offchain

Reply:
Because we'd like to guarantee that all the parameters provided in the function contribute to the entropy of the addresses. This is a nice abstraction that is very useful when composing factories.

Reply:
What we want, instead, is that it won't matter who triggers the deterministic deployment. The final contract deployed is the same, as it is fully specified by the arguments provided to the factory function. Now, because `authorityOwner` is not forwarded to the `AuthorityFactory` contract, it wouldn't participate in the entropy of calculating the address of the `Authority` contract. That is why we force it to participate in the entropy by generating a new salt, composed by it and the "old" salt.

Reply:
It's a nice property to have, and it's cheap to implement. :-)